retrofit-mock/src/main/java/retrofit2/mock/BehaviorDelegate.java

```diff
@@ -54,11 +60,91 @@ public T returningResponse(@Nullable Object response) {
-          return callAdapter.adapt(behaviorCall);
+              (CallAdapter<R, T>) retrofit.callAdapter(adapterInfo.responseType, methodAnnotations);
+
+          T adapted = callAdapter.adapt(behaviorCall);
+          if (!adapterInfo.isSuspend) {
+            return adapted;
+          }
+
+          Call<Object> adaptedCall = (Call<Object>) adapted;
+          Continuation<Object> continuation = (Continuation<Object>) args[args.length - 1];
+          try {
+            if (adapterInfo.wantsResponse) {
+              return KotlinExtensions.awaitResponse(adaptedCall, continuation);
+            }
+            return KotlinExtensions.await(adaptedCall, continuation);
+          } catch (Exception e) {
+            return KotlinExtensions.suspendAndThrow(e, continuation);
+          }
         });
   }
+
+  private static ServiceMethodAdpaterInfo parseServiceMethodAdapterInfo(Method method) {
+    Type[] genericParameterTypes = method.getGenericParameterTypes();
+    if (genericParameterTypes.length != 0) {
+      Type lastParameterType = genericParameterTypes[genericParameterTypes.length - 1];
+      if (lastParameterType instanceof ParameterizedType) {
+        ParameterizedType parameterizedLastParameterType = (ParameterizedType) lastParameterType;
+        try {
+          if (parameterizedLastParameterType.getRawType() == Continuation.class) {
+            Type resultType = parameterizedLastParameterType.getActualTypeArguments()[0];
+            if (resultType instanceof WildcardType) {
+              resultType = ((WildcardType) resultType).getLowerBounds()[0];
+            }
+            if (resultType instanceof ParameterizedType) {
+              ParameterizedType parameterizedResultType = (ParameterizedType) resultType;
+              if (parameterizedResultType.getRawType() == Response.class) {
+                Type bodyType = parameterizedResultType.getActualTypeArguments()[0];
+                Type callType = new CallParameterizedTypeImpl(bodyType);
+                return new ServiceMethodAdpaterInfo(true, true, callType);
+              }
+            }
+            Type callType = new CallParameterizedTypeImpl(resultType);
+            return new ServiceMethodAdpaterInfo(true, false, callType);
+          }
+        } catch (NoClassDefFoundError ignored) {
+          // Not using coroutines.
+        }
+      }
+    }
+    return new ServiceMethodAdpaterInfo(false, false, method.getGenericReturnType());
+  }
+
+  static final class CallParameterizedTypeImpl implements ParameterizedType {
+    private final Type bodyType;
+
+    CallParameterizedTypeImpl(Type bodyType) {
+      this.bodyType = bodyType;
+    }
+
+    @Override public Type[] getActualTypeArguments() {
+      return new Type[] { bodyType };
+    }
+
+    @Override public Type getRawType() {
+      return Call.class;
+    }
+
+    @Override public @Nullable Type getOwnerType() {
+      return null;
+    }
+  }
+
+  static class ServiceMethodAdpaterInfo {
```

Comment:
ubernit: `Adapter`

Reply:
Can't see typos in this new font @saket...

Reply:
That's exactly what I say when @AlecStrong or @oldergod find bugs in my PRs